usr/lib/api/api_interface.c

```diff
@@ -2960,6 +2961,37 @@ CK_RV C_Initialize(CK_VOID_PTR pVoid)
         return CKR_CANT_LOCK;
     }
 
+    /*
+     * Ensure OpenSSL is initialized and the OpenSSL config is loaded
+     * at the very beginning of C_Initialize(). Loading the OpenSSL config
+     * may cause engines or providers to be loaded, which in turn may initialize
+     * Opencryptoki again, if an engine or provider calls C_Initialize as
+     * part of its initialization.
+     * Special handling is required to allow this recursive use of C_Initialize
+     * to avoid a CKR_CRYPTOKI_ALREADY_INITIALIZED error.
+     * OPENSSL_init_crypto must be called without holding the GlobMutex as this
+     * would cause a deadlock for such recursive calls of C_Initialize.
+     */
+    before_anchor = Anchor;
+    pthread_mutex_unlock(&GlobMutex);
```

Comment:
Hm, yes, the unlock at the end of C_Initialize is also without error checking. The question what should we do if it fails? If it was successfully locked before, and unlock fails it will stay locked, and probably causing a deadlock later on. All we could do is to issue a trace message, but unfortunately the trace has not been set up at line 2976, so tracing most likely won't work either.... 
Any yes, the trace for the lock-error case in line 2960 above will also not work :-)

Reply:
The man page lists the possible error cases:
```
The pthread_mutex_lock(), pthread_mutex_trylock(), and pthread_mutex_unlock() functions may fail if:

EINVAL
    The value specified by mutex does not refer to an initialized mutex object. 
EAGAIN
    The mutex could not be acquired because the maximum number of recursive locks for mutex has been exceeded.

[...]

The pthread_mutex_unlock() function may fail if:

EPERM
    The current thread does not own the mutex.
```

- EINVAL: this case is very unlikely for unlock, if the lock was successful. It can IMHO only happen in case of a coding error.
- EPERM: same for this one.
- EAGAIN: this is the only case, that we possibly should handle by yield and re-try again.

Reply:
The pthread_mutex_lock calls are error checked, only the pthread_mutex_unlock is is not. 
With that, I would say it is save to not check errors on pthread_mutex_unlock.

Reply:
Hm, but does pthread_mutex_unlock really return EAGAIN? Its about acquiring the lock, which is not what unlock does, no? Also this is no recursive lock.

Reply:
I agree, it's a bit weird that EAGAIN is listed for unlock. I can't find any path in the libpthread code, which returns EAGAIN for unlock. So maybe it's a bug in the documentation. Maybe we can check for the return value and log it, if it is != 0. I think, this should be sufficient.

Reply:
The trace environment has not been set up at that point of the code, so adding a trace statement would just do nothing.... And adding a fprintf to stderr is not what I want to do in a library....
So I would tend to leave it as is.